yksgi/src/lib.rs

```diff
@@ -0,0 +1,357 @@
+use llvm_sys::core::*;
+use llvm_sys::{LLVMOpcode, LLVMTypeKind};
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::ptr;
+
+mod llvmbridge;
+use llvmbridge::{Module, BasicBlock, Value, Type, get_aot_original};
+
+static ALWAYS_SKIP_FUNCS: [&str; 3] = ["llvm.lifetime", "llvm.lifetime.end.p0i8", "llvm.dbg.value"];
+static SKIP_FOR_NOW_FUNCS: [&str; 2] = ["yk_location_drop", "yk_mt_drop"];
+
+/// Stopgap interpreter values.
+#[derive(Clone, Copy, PartialEq)]
+pub struct SGValue {
+    pub val: u64,
+    pub ty: Type,
+}
+
+impl SGValue {
+    pub fn new(val: u64, ty: Type) -> Self {
+        SGValue { val, ty }
+    }
+
+    pub fn with_type(&self, ty: Type) -> Self {
+        SGValue::new(self.val, ty)
+    }
+
+    pub fn as_str(&self) -> String {
+        format!("SGValue ( {}, {:?} )", self.val, self.ty.as_str())
+    }
+}
+
+/// A frame holding live variables.
+struct Frame {
+    vars: HashMap<Value, SGValue>,
+}
+
+impl Frame {
+    fn new() -> Frame {
+        Frame {
+            vars: HashMap::new(),
+        }
+    }
+
+    /// Get the value of the variable `key` in this frame.
+    fn get(&self, key: &Value) -> Option<&SGValue> {
+        self.vars.get(key)
+    }
+
+    /// Add new variable `key` with value `val`.
+    fn add(&mut self, key: Value, val: SGValue) {
+        debug_assert!(!self.vars.contains_key(&key));
+        self.vars.insert(key, val);
+    }
+}
+
+/// The stopgap interpreter. Used during guard failures to get back to the control point by
+/// interpreting LLVM IR.
+pub struct SGInterp {
+    /// LLVM IR module we are interpreting.
+    module: Module,
+    /// Current frames.
+    frames: Vec<Frame>,
+    /// Current instruction being interpreted.
+    pc: Value,
```

Comment:
We seem to call `unwrap` on this `Value` quite a lot, but I can't work out why. Any hints?